File: CS/WinExample/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraEditors.DXErrorProvider;
using DevExpress.Persistent.AuditTrail;
using System.Reflection;
using System.Security.Principal;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Xpo;

namespace WinExample {
    public partial class Form1 : Form {
        public Form1() {
            XpoDefault.DataLayer = XpoDefault.GetDataLayer("XpoProvider=InMemoryDataStore", DevExpress.Xpo.DB.AutoCreateOption.DatabaseAndSchema);
            InitializeComponent();
        }
        private void Form1_Load(object sender, EventArgs e) {
            AuditTrailService.Instance.QueryCurrentUserName += new QueryCurrentUserNameEventHandler(Instance_QueryCurrentUserName);
            session1.Dictionary.GetDataStoreSchema(typeof(PersistentObject1).Assembly);
            AuditTrailService.Instance.SetXPDictionary(session1.Dictionary);
            AuditTrailService.Instance.AuditDataStore = new AuditDataStore<AuditDataItemPersistent, AuditedObjectWeakReference>();
            AuditTrailService.Instance.BeginSessionAudit(session1, AuditTrailStrategy.OnObjectChanged, ObjectAuditingMode.Full);
        }
        void Instance_QueryCurrentUserName(object sender, QueryCurrentUserNameEventArgs e) {
            e.CurrentUserName = WindowsIdentity.GetCurrent().Name;
        }
        private void Form1_FormClosed(object sender, FormClosedEventArgs e) {
            AuditTrailService.Instance.SaveAuditData(session1);
        }
    }
}

File: VB/WinExample/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraEditors.DXErrorProvider
Imports DevExpress.Persistent.AuditTrail
Imports System.Reflection
Imports System.Security.Principal
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Xpo

Namespace WinExample
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			XpoDefault.DataLayer = XpoDefault.GetDataLayer("XpoProvider=InMemoryDataStore", DB.AutoCreateOption.DatabaseAndSchema)
			InitializeComponent()
		End Sub
		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			AddHandler AuditTrailService.Instance.QueryCurrentUserName, AddressOf Instance_QueryCurrentUserName
			session1.Dictionary.GetDataStoreSchema(GetType(PersistentObject1).Assembly)
			AuditTrailService.Instance.SetXPDictionary(session1.Dictionary)
			AuditTrailService.Instance.AuditDataStore = New AuditDataStore(Of AuditDataItemPersistent, AuditedObjectWeakReference)()
			AuditTrailService.Instance.BeginSessionAudit(session1, AuditTrailStrategy.OnObjectChanged, ObjectAuditingMode.Full)
		End Sub
		Private Sub Instance_QueryCurrentUserName(ByVal sender As Object, ByVal e As QueryCurrentUserNameEventArgs)
			e.CurrentUserName = WindowsIdentity.GetCurrent().Name
		End Sub
		Private Sub Form1_FormClosed(ByVal sender As Object, ByVal e As FormClosedEventArgs) Handles MyBase.FormClosed
			AuditTrailService.Instance.SaveAuditData(session1)
		End Sub
	End Class
End Namespace
